### PR TITLE
Bach BQ: support full table names in `from_table()`

### PR DESCRIPTION
**Pull request by thijs-obj**
This solves issue https://github.com/objectiv/objectiv-analytics/issues/992

Change:
* Support full table names for BigQuery in `from_table()`, that is `project_id.dataset.table_name` is now supproted
* Remove `bq_dataset` and `bq_project_id` parameters. I don't think they were every used anywhere. And I highly doubt any users depend on it. So seems safe to remove

Reason for this change is more consistency, between what we allow in `database_create_table()`, `get_sample()`, and now `from_table()`. Resolves this discussion in a recent PR: https://github.com/objectiv/objectiv-analytics/pull/976/files#r914061937

### CODE REVIEW

File: bach/bach/from_database.py [hunk 1]
```diff
@@ -71,6 +65,25 @@ def get_dtypes_from_table(
...
+    """ Split a BigQuery table name into project, dataset and actual table_name. """
+    parts = table_name.split('.')
+    if len(parts) == 3:
+        return FullTableName(project_id=parts[0], dataset=parts[1], table_name=parts[2])
+    if len(parts) == 2:
+        return FullTableName(project_id=None, dataset=parts[0], table_name=parts[1])
+    if len(parts) == 1:
+        return FullTableName(project_id=None, dataset=None, table_name=parts[0])
+    raise ValueError(f'Expected table_name in format "{{project_id}}.{{dataset}}.{{table_name}}"'
+                     f', value: {table_name}')
```
Comment (KathiaBarahona) [inline bach/bach/from_database.py:84]: mmm  why not just having a regex? smth like `r'^([^.]+\.){0,2}(?P<table_name>[^.]+)$'`. Seems to me we just need to check if the string is in the correct format and just replace the string from the `table_name` group with `INFORMATION_SCHEMA.COLUMNS`.

File: bach/bach/dataframe.py [hunk 2]
```diff
@@ -597,7 +575,7 @@ def from_table(
         }
         sql_params.update(column_placeholders)
 
-        sql = f'SELECT {column_stmt} FROM {sql_table_name_template}'
+        sql = f'SELECT {column_stmt} FROM {{table_name}}'
```
Comment (KathiaBarahona) [inline bach/bach/dataframe.py:578]: **question**: shouldn't we check the `table_name` is in the right format for BQ? Else user will only know they provided the wrong table_name till they try to execute a query (if they actually provided the dtypes)

File: bach/bach/from_database.py [hunk 3]
```diff
@@ -71,6 +65,25 @@ def get_dtypes_from_table(
     return _get_dtypes_from_information_schema_query(engine=engine, query=sql)
 
 
+class FullTableName(NamedTuple):
+    project_id: Optional[str]
+    dataset: Optional[str]
+    table_name: str
+
+
+def _split_bq_table_name(table_name: str) -> FullTableName:
```
Comment (KathiaBarahona) [inline bach/bach/from_database.py:74]: Can we add a test for this?

**Review by KathiaBarahona — COMMENTED** [review]
Some comments

File: bach/bach/dataframe.py (hunk 2 above)
Comment (thijs-obj) [reply bach/bach/dataframe.py:578]: In general, if a user gives a non-existing table name, then we won't know that that table doesn't exist until we try to query it.

We could check the specific sub-case where the table is not only non-existent but also the name would never be allowed as a table name. But I am doubting a bit how much extra worth that would have to the user.

File: bach/bach/from_database.py (hunk 1 above)
Comment (thijs-obj) [reply bach/bach/from_database.py:84]: Well, I thought having `split('.')` was simpler than having a regex, and the result from split is somewhat similar as what it was before: three separate values, two of which could be None. This made the code change in `get_dtypes_from_table()` sort of smaller.

We don't just need to replace `table_name` in the regex with `INFORMATION_SCHEMA.COLUMNS`, we also need the `table_name` value itself

File: bach/bach/from_database.py [hunk 4]
```diff
@@ -35,31 +35,25 @@ def get_dtypes_from_model(engine: Engine, node: SqlModel) -> Dict[str, Structure
...
-    bq_dataset: Optional[str] = None,
-    bq_project_id: Optional[str] = None
 ) -> Dict[str, StructuredDtype]:
     """
     Query database to get dtypes of the given table.
     :param engine: sqlalchemy engine for the database.
-    :param table_name: the table name for which to get the dtypes
-    :param bq_dataset: BigQuery-only. Dataset in which the table resides, if different from engine.url
-    :param bq_project_id: BigQuery-only. Project of dataset, if different from engine.url
+    :param table_name: the table name for which to get the dtypes.
```
[the remark this answers is not in the document]
Comment (thijs-obj) [reply bach/bach/from_database.py]: doh. Fixed in https://github.com/objectiv/objectiv-analytics/pull/998/commits/93fd8777e906fa96d474f261bc685a381f6235ba

File: bach/bach/from_database.py (hunk 1 above)
Comment (thijs-obj) [reply bach/bach/from_database.py]: I further simplified and let go of splitting this in three parts. Split in two now. And refactored it all a bit in https://github.com/objectiv/objectiv-analytics/pull/998/commits/93fd8777e906fa96d474f261bc685a381f6235ba

let me know what you think

**Pull request merged**